2019-08-03-q-and-as-about-nix-for-haskellers.md

```diff
@@ -0,0 +1,692 @@
+
+```nix
+mkDerivation {
+  pname = "conduit";
+  version = "1.3.1.1";
+  sha256 = "84dfafc92e9553c7bae4b4fe0cba3da29b37def606f88b989db95ee2dc933fa2";
+  libraryHaskellDepends = [
+    base bytestring directory exceptions filepath mono-traversable mtl
+    primitive resourcet text transformers unix unliftio-core vector
+  ];
+  doHaddock = false;
+  doCheck = false;
+  homepage = "http://github.com/snoyberg/conduit";
+  description = "Streaming data processing library";
+  license = stdenv.lib.licenses.mit;
+}
+```
+
+This is a derivation for building `conduit`.  You can see that it just
+duplicates some of the info from the `.cabal` file.[^5]
+
+nixpkgs provides a set of these Haskell derivations for building all the
+Haskell packages from Hackage.  nixpkgs provides easy ways of injecting your
+own packages into this package set, or changing the versions of packages in
+this package set (you can see above that the `conduit` version is fixed to
+version 1.3.1.1).  nixpkgs also provides a way to define your own custom
+Haskell package set.
+
+The Haskell package set is used in different ways depending on whether you're
+using the Nix CLI tools, using `cabal` + Nix, or using `stack` + Nix.  I
+describe each of these methods below:
+
+1.  _Use `nix-build` from the CLI to build Haskell projects._
+
+    This method doesn't directly involve `cabal` or `stack`.[^6]
+
+    In this method, the developer generates a derivation corresponding to the
+    Haskell package they are developing.  They then call `nix-build` on this
+    derivation.  When `nix-build` tries to build the derivation, it gets
+    Haskell dependencies from the nixpkgs Haskell package set.  It also gets
+    system dependencies (like `zlib` or PostgreSQL bindings) from nixpkgs.
+
+    `nix-build` is not good for interative development, since it doesn't know
+    how to only rebuild parts of your code that have changed.  It does a full
+    rebuild every time it is run.
+
+    `nix-build` is mainly used in two different situations.  One is by
+    end-users who are trying to install an exectuable written in Haskell.
+    For instance, if you want to try out the terminal emulator
+    [Termonad](https://github.com/cdepillabout/termonad),
+    you can just clone the repository and run `nix-build`.  You don't have to
+    worry about having `stack` or `cabal` on your system, or making sure any of
+    the GTK libraries are installed.  `nix-build` handles everything for you.
+
+    The other situation is in CI.  Using `nix-build` in CI gives extremely high
+    confidence that your build is actually reproducible.  In practice this ends
+    up being much nicer than directly calling `stack` or `cabal`.
+
+2.  _Use `cabal` + Nix to build Haskell projects._
+
+    This method uses a combination of `cabal` and Nix.
+
+    In this method, Nix is used to build a GHC package database of all your
+    Haskell dependencies (as well as their required system packages).
```

Comment:
It's not clear what is GHC package database and how it's different from Haskell package database or something else :thinking: